// arco-markdown-loader/lib/utils/parseMarkdownAttributes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const front_matter_1 = __importDefault(require("front-matter"));
const marked_1 = __importDefault(require("../parser/marked"));
function generateHeaderHtml(header) {
    return header
        ? `<div className="ac-nav-intro">${(0, marked_1.default)(header.replace(/(\/) (.*)/, '<span className="separator">$1</span><strong>$2</strong>'))}</div>`
        : '';
}
function trimHeaderDefinedInMarkdownContent(markdown, formatTitle) {
    var _a, _b, _c;
    const header = ((_a = /^(([ \t]*`{5})([^\n]*)([\s\S]+?)(^[ \t]*\2))/m.exec(markdown)) === null || _a === void 0 ? void 0 : _a[4]) || '';
    const title = ((_b = /# (.*)/.exec(header)) === null || _b === void 0 ? void 0 : _b[1]) || '';
    return {
        headerHtml: generateHeaderHtml(header),
        title: typeof formatTitle === 'function' ? formatTitle(title) : title,
        description: ((_c = /# .*\n*(.*)/.exec(header)) === null || _c === void 0 ? void 0 : _c[1]) || '',
        markdown: markdown.replace(`\`\`\`\`\`${header}\`\`\`\`\``, ''),
    };
}
function parseMarkdownAttributes(originalMarkdown, formatTitle) {
    const { body: markdownBody, attributes } = (0, front_matter_1.default)(originalMarkdown);
    const trimResult = trimHeaderDefinedInMarkdownContent(markdownBody, formatTitle);
    /**
     * Get markdown header from content like below
     * `````
     * Group
     * # Title
     * Some description...
     * `````
     */
    if (trimResult.headerHtml) {
        return Object.assign(Object.assign({}, trimResult), { attributes });
    }
    /**
     * Get markdown header from front matter like below
     * ---
     * group: Group
     * title: Title
     * description: Some description...
     * ---
     */
    const { title = '', description = '', group = '' } = attributes;
    return {
        title,
        description,
        attributes,
        markdown: trimResult.markdown,
        headerHtml: title || description || group
            ? generateHeaderHtml([group, title.replace(/^([^#])/, '# $1'), description].join('\n'))
            : '',
    };
}
exports.default = parseMarkdownAttributes;
