// @arco-materials/material-preview-utils/es/constant.js
// 跨窗口通信
export var MESSAGE_KEY = 'arcoMaterialMessage';
// 物料预览页面初始化完毕
export var MESSAGE_TYPE_PAGE_LOADED = 'page-loaded';
// 物料加载完毕
export var MESSAGE_TYPE_MATERIAL_LOADED = 'material-loaded';
// 物料 CodeBlock 更新
export var MESSAGE_TYPE_CODE_UPDATE = 'code-update';
// 父页面准备就绪，允许子页面进行物料渲染
export var MESSAGE_TYPE_PARENT_WINDOW_READY = 'parent-window-ready';
// 重新加载 iframe 中的子页面
export var MESSAGE_TYPE_CONTENT_WINDOW_RELOAD = 'content-window-reload';
// 切换 Arco 亮暗主题的消息
export var MESSAGE_TYPE_ARCO_THEME_CHANGE = 'arco-theme-change';
// Arco React 包信息
export var ARCO_REACT_PACKAGE_NAME = {
    v1: '@bytedesign/web-react',
    v2: '@arco-design/web-react',
};
export var ARCO_REACT_FILENAME_JS = {
    v1: 'byteui.min.js',
    v2: 'arco.min.js',
};
export var ARCO_REACT_FILENAME_CSS = {
    v1: 'byteui.css',
    v2: 'arco.css',
};
export var ARCO_REACT_FILENAME_ICON = {
    v1: 'byteui-icon.min.js',
    v2: 'arco-icon.min.js',
};
export var ARCO_REACT_MODULE_NAME = {
    v1: 'byteui',
    v2: 'arco',
};
export var ARCO_REACT_MODULE_NAME_ICON = {
    v1: 'byteicon',
    v2: 'arcoicon',
};
// Arco Vue 包信息
export var ARCO_VUE_PACKAGE_NAME = '@arco-design/web-vue';
export var ARCO_VUE_FILENAME_JS = 'arco-vue.js';
export var ARCO_VUE_FILENAME_CSS = 'arco.css';
export var ARCO_VUE_FILENAME_ICON = 'arco-vue-icon.js';
export var ARCO_VUE_MODULE_NAME = 'ArcoVue';
export var ARCO_VUE_MODULE_NAME_ICON = 'ArcoVueIcon';
// 物料包信息
export var MATERIAL_FILENAME_JS = 'arco.min.js';
export var MATERIAL_FILENAME_VENDOR = 'arcoDemoVendor.min.js';
export var MATERIAL_FILENAME_CSS = 'index.css';
export var MATERIAL_MODULE_NAME = 'arco';
export var MATERIAL_MODULE_NAME_VENDOR = 'arcoDemoVendor';
export var REGEXP_IMPORT_ARCO_REACT = /^@(arco-|byte)design\/web-react(\/(es|lib))?/;
export var REGEXP_IMPORT_ARCO_REACT_ICON = /^@(arco-|byte)design\/web-react\/icon$/;
export var REGEXP_IMPORT_ARCO_VUE = /^@arco-design\/web-vue(\/(es|lib))?/;
export var REGEXP_IMPORT_ARCO_VUE_ICON = /^@arco-design\/web-vue\/es\/icon$/;
export var REGEXP_IMPORT_MATERIAL_DEMO_VENDOR = /\.+\/arcoDemoVendor/;
export var LIBRARY_GLOBAL_NAME_MAP = {
    lodash: '_',
    vue: 'Vue',
    react: 'React',
    'react-dom': 'ReactDOM',
    'react-router-dom': 'ReactRouterDOM',
    'lodash-es': '_',
    'date-fns': 'dateFns',
    arcoLibrary: 'arcoLibrary',
    arcoMaterial: 'arcoMaterial',
    arcoMaterialDemoVendor: 'arcoMaterialDemoVendor',
};
// 用于自动截图的类名
export var CLASSNAME_FOR_AUTO_SCREENSHOT = 'arco-material-live-preview';
// 本地预览物料时的 URL Query 参数
export var QUERY_LOCAL_PREVIEW_URL = 'localPreviewUrl';
// 无需注入 Arco 的版本号
export var ARCO_VERSION_NONE = 'none';
