// @arco-materials/material-preview-utils/es/injectGlobalScope.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React from 'react';
import ReactDOM from 'react-dom';
import * as dateFns from 'date-fns';
import * as _ from 'lodash';
import * as ReactRouterDOM from 'react-router-dom';
export var GLOBALS = {
    React: React,
    ReactDOM: ReactDOM,
    ReactRouterDOM: ReactRouterDOM,
    _: _,
    lodash: _,
    dateFns: dateFns,
    'process.env.NODE_ENV': process.env.NODE_ENV,
};
export var GLOBAL_MODULE_MAP = {
    react: 'React',
    'react-dom': 'ReactDOM',
    'react-router-dom': 'ReactRouterDOM',
    lodash: 'lodash',
    'date-fns': 'dateFns',
};
/**
 * 向全局插入以下变量 React, ReactDOM, ReactRouterDOM, _, lodash, dateFns
 */
export default function injectGlobalScope(window, globals) {
    return _.each(__assign(__assign({}, GLOBALS), globals), function (val, key) {
        _.set(window, key, val);
    });
}
